src/NServiceBus.Core/Routing/UnicastSubscriberTable.cs

```diff
@@ -84,9 +82,8 @@ public void AddOrReplaceRoutes(string sourceKey, IList<RouteTableEntry> entries)
             return newRouteTable.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToArray());
         }
 
-        Dictionary<Type, UnicastRoute[]> routeTable = new Dictionary<Type, UnicastRoute[]>();
-        ConcurrentDictionary<object, IList<RouteTableEntry>> routeGroups = new ConcurrentDictionary<object, IList<RouteTableEntry>>();
-        long enterCounter;
-        long exitCounter = 1;
+        volatile Dictionary<Type, UnicastRoute[]> routeTable = new Dictionary<Type, UnicastRoute[]>();
+        Dictionary<string, IList<RouteTableEntry>> routeGroups = new Dictionary<string, IList<RouteTableEntry>>();
+        ReaderWriterLockSlim @lock = new ReaderWriterLockSlim();
```

Comment:
I don't like explicit names like this for sync primitives, but that's core area so I'll apply it.